xapian-letor/tests/api_letor.cc

```diff
@@ -866,10 +869,10 @@ DEFINE_TESTCASE(colltfcolllenfeature, generated)
     Xapian::CollTfCollLenFeature feature;
 
     Xapian::Database db(db_path);
-    feature.set_database(db);
+    feature.internal->set_database(db);
 
     Xapian::Document doc = db.get_document(1);
-    feature.set_doc(doc);
+    feature.internal->set_doc(doc);
```

Comment:
Oh, except those setters are used here...

But I think if we're trying to test a particular feature we should do that via a `FeatureList` holding it, and test the reported values of the feature.

While if we're wanting to test the stats that features see, we should instead create a custom `Feature` subclass for that purpose (like `CheckStatsWeight` in `xapian-core/tests/api_weight.cc` which is used to test the stats provided to `Weight` subclasses), and again test via a `FeatureList` holding it.

Currently it looks to me like we seem to be trying to do both at the same time here.

Reply:
> But I think if we're trying to test a particular feature we should do that via a `FeatureList` holding it, and test the reported values of the feature.
> 
This tests are unittests which means we can't use any other api's for testing them and I have already written tests which check a particular feature using FeatureList in #252(checkcreatefeaturevector_(feature_name)).
> While if we're wanting to test the stats that features see, we should instead create a custom `Feature` subclass for that purpose (like `CheckStatsWeight` in `xapian-core/tests/api_weight.cc` which is used to test the stats provided to `Weight` subclasses), and again test via a `FeatureList` holding it.
>
I think there is no need for this as in case of `Weight` class the statistics were protected and hence there was a need for testing them via custom subclass but here internal is public and we can easily access the maps via getter and setter methods.

Reply:
The distinction between unittest and not is somewhat arbitrary. It's absolutely fine to test via another class — particularly if that's the way they are used in the codebase.

Reply:
Very few testcases in our testsuite are unit tests of a single class, but that's indeed really not a problem.

In `core`, `unittest` and `internaltest` unit test things low-level functionality inside the library by linking internal code directly into the test program.  These tests cover things which are hard to reliably test via the public API.  Everything else is tested via the public APIs which users use.

In some cases a public API class can be tested in isolation (e.g. `stemtest` has unit tests of the `Xapian::Stem` API), but generally API classes interact.

Currently `api_letor.cc` has `#include <api/feature_internal.h>` which really isn't something that should be included there.  The `internal` member currently has `public` visibility, but it's definitely not intended for public use.

Reply:
fixed in a fixup commit.